File: examples/15_algorithms/15d_segment_tree/segment_tree.c
/* The same iterative segment tree as segment_tree.ppy, same workload. */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static long long build(long long *tree, long long size) {
    for (long long i = size - 1; i > 0; i--) {
        tree[i] = tree[2 * i] + tree[2 * i + 1];
    }
    return tree[1];
}

static long long update(long long *tree, long long size, long long position, long long value) {
    long long index = position + size;
    tree[index] = value;
    index /= 2;
    while (index >= 1) {
        tree[index] = tree[2 * index] + tree[2 * index + 1];
        index /= 2;
    }
    return tree[1];
}

static long long query(const long long *tree, long long size, long long left, long long right) {
    long long total = 0;
    long long low = left + size;
    long long high = right + size;
    while (low < high) {
        if (low % 2 == 1) {
            total += tree[low];
            low += 1;
        }
        if (high % 2 == 1) {
            high -= 1;
            total += tree[high];
        }
        low /= 2;
        high /= 2;
    }
    return total;
}

static long long workload(long long *tree, long long size, long long rounds) {
    long long checksum = 0;
    for (long long step = 0; step < rounds; step++) {
        long long position = (step * 7919) % size;
        update(tree, size, position, (step * 31) % 1000);
        long long left = (step * 104729) % size;
        long long right = left + (step % 512) + 1;
        right = right > size ? size : right;
        checksum += query(tree, size, left, right);
    }
    return checksum;
}

int main(void) {
    const long long size = 1 << 18;
    long long *tree = calloc((size_t)(2 * size), sizeof(long long));
    for (long long i = 0; i < size; i++) {
        tree[size + i] = (i * 7919) % 1000;
    }
    build(tree, size);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    long long checksum = workload(tree, size, 200000);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double ms = (t1.tv_sec - t0.tv_sec) * 1000.0 + (t1.tv_nsec - t0.tv_nsec) / 1e6;
    printf("segment tree 2e5%8.1f ms   -> %lld\n", ms, checksum);
    return 0;
}

File: examples/15_algorithms/15d_segment_tree/segment_tree.py
"""Range sums with point updates, over a segment tree.

Input:  N M K on the first line, then N numbers, then M+K command lines --
        `1 b c` assigns c to position b, `2 b c` asks for the sum over
        [b, c).
Output: the checksum of every answered query, so one line stands in for K.
"""

import array


def build(tree, size):
    for i in range(size - 1, 0, -1):
        tree[i] = tree[2 * i] + tree[2 * i + 1]
    return tree[1]


def update(tree, size, position, value):
    index = position + size
    tree[index] = value
    index //= 2
    while index >= 1:
        tree[index] = tree[2 * index] + tree[2 * index + 1]
        index //= 2
    return tree[1]


def query(tree, size, left, right):
    total = 0
    low = left + size
    high = right + size
    while low < high:
        if low % 2 == 1:
            total += tree[low]
            low += 1
        if high % 2 == 1:
            high -= 1
            total += tree[high]
        low //= 2
        high //= 2
    return total


def run_commands(tree, commands, size, rounds):
    checksum = 0
    for step in range(rounds):
        base = step * 6
        update(tree, size, commands[base + 1], commands[base + 2])
        checksum += query(tree, size, commands[base + 4], commands[base + 5])
    return checksum


def generated():
    """The stand-in input, in the same order the judge would send it."""
    size = 1 << 18
    rounds = 200000
    fields = [size, rounds, rounds]
    fields.extend((i * 7919) % 1000 for i in range(size))
    for step in range(rounds):
        fields.append(1)
        fields.append((step * 7919) % size)
        fields.append((step * 31) % 1000)
        left = (step * 104729) % size
        fields.append(2)
        fields.append(left)
        fields.append(min(left + (step % 512) + 1, size))
    return fields


def read_fields():
    """The judge's input, or a generated stand-in when nothing is piped in."""
    try:
        size, rounds, queries = map(int, input().split())
    except EOFError:
        return array.array("q", generated())
    fields = array.array("q", [size, rounds, queries])
    rest = array.array("q", [0] * (size + rounds * 6))
    for i in range(size + rounds * 6):
        rest[i] = int(input())
    fields.extend(rest)
    return fields


def main():
    fields = read_fields()
    size = fields[0]
    rounds = fields[1]
    tree = array.array("q", [0] * (2 * size))
    for i in range(size):
        tree[size + i] = fields[3 + i]
    commands = array.array("q", fields[3 + size :])

    build(tree, size)
    print(run_commands(tree, commands, size, rounds))


main()
